packages/react-dom/src/test-utils/ReactTestUtils.js

```diff
@@ -390,58 +438,120 @@ const ReactTestUtils = {
     }
-
-    const result = ReactDOM.unstable_batchedUpdates(callback);
     // note: keep these warning messages in sync with
     // createReactNoop.js and ReactTestRenderer.js
-    if (__DEV__) {
-      if (result !== undefined) {
-        let addendum;
-        if (result !== null && typeof result.then === 'function') {
-          addendum =
-            '\n\nIt looks like you wrote ReactTestUtils.act(async () => ...), ' +
-            'or returned a Promise from the callback passed to it. ' +
-            'Putting asynchronous logic inside ReactTestUtils.act(...) is not supported.\n';
-        } else {
-          addendum = ' You returned: ' + result;
+    let thenable;
+    actedUpdates((doesHavePassiveEffects, onDone) => {
+      const result = ReactDOM.unstable_batchedUpdates(callback);
+      if (
+        result !== null &&
+        typeof result === 'object' &&
+        typeof result.then === 'function'
+      ) {
+        // setup a boolean that gets set to true only
+        // once this act() call is await-ed
+        let called = false;
+        if (__DEV__) {
+          if (typeof Promise !== 'undefined') {
+            //eslint-disable-next-line no-undef
+            Promise.resolve()
+              .then(() => {})
+              .then(() => {
+                if (called === false) {
+                  warningWithoutStack(
+                    null,
+                    'You called act(async () => ...) without await. ' +
+                      'This could lead to unexpected testing behaviour, interleaving multiple act ' +
+                      'calls and mixing their scopes. You should - await act(async () => ...);',
+                  );
+                }
+              });
+          }
         }
-        warningWithoutStack(
-          false,
-          'The callback passed to ReactTestUtils.act(...) function must not return anything.%s',
-          addendum,
-        );
-      }
-    }
-    ReactDOM.render(<div />, actContainerElement);
-    // we want the user to not expect a return,
-    // but we want to warn if they use it like they can await on it.
-    return {
-      then() {
+
+        // in this case, the returned thenable runs the callback, flushes
+        // effects and  microtasks in a loop until doesHavePassiveEffects() === false,
+        // and cleans up
+        thenable = {
+          then(successFn, errorFn) {
+            called = true;
+            result.then(
+              () => {
+                flushEffectsAndMicroTasks(doesHavePassiveEffects, () => {
+                  if (onDone !== undefined) {
+                    onDone();
+                  }
+                  if (typeof successFn === 'function') {
+                    successFn();
+                  }
+                });
+              },
+              err => {
+                if (onDone !== undefined) {
+                  onDone(err);
+                }
+                errorFn(err);
```

Comment:
Unlike with `successFn` above, you're not checking if `errorFn` exists. That's probably ok, because we expect people to use async/await, or wrap in Promise.resolve. But it should be consistent regardless.

Reply:
Ah good catch. Will fix both. I think I had a test where I’d manually called .then. I’ll remove the checks.